src/python/pants/backend/python/targets/python_target.py

```diff
@@ -21,7 +20,14 @@ class PythonTarget(Target):
     """
 
     def __init__(
-        self, address=None, payload=None, sources=None, provides=None, compatibility=None, **kwargs
+        self,
+        address=None,
+        payload=None,
+        sources=None,
+        provides=None,
+        compatibility=None,
+        constraints=None,
```

Comment:
Shouldn't constraints be limited to binaries? It's only binaries that you generally want to pin with a lock file. Fo example, if we allow a library to pick constraints, what do we do with the dist it exports if it does export a dist?

Reply:
Good question re the dist. I would guess we ignore the constraints for that purpose? If you want to lock things down at the formal dep level , do so in requirements.txt.

Reply:
But more generally I think that since a global constraints file would apply to libraries, they may as well have local constraints too. I can see it making sense to say "even though our global requirement on foo is >=2.0 I happen to know that this specific library has code in it that requires >=2.1". It makes more sense for that to live adjacent to the code that imposes the constraint.

Reply:
> "even though our global requirement on foo is >=2.0 I happen to know that this specific library has code in it that requires >=2.1".

This example seems particularly likely to me in the case of a) huge repositories like Twitter and b) multirepo users migrating to a monorepo. It seems very likely that someone migrating from a multirepo will need to be able to say `these 10 targets, use ==2.2; these 4 targets, use ==2.3`.

Reply:
> "even though our global requirement on foo is >=2.0 I happen to know that this specific library has code in it that requires >=2.1". It makes more sense for that to live adjacent to the code that imposes the constraint.

That I agree with - but that is not a constraint, its a requirement. It's clunky today to write a custom requirement in the same BUILD file because you must write a `python_requirement_library`, but you might remember the dependency list used to be heterogenous in Pants v0 allowing the spiritual equivalent of (can't remember exact syntax):
```python
python_binary(
  name='app',
  dependencies=[
    '3rdparty/python:pex',
    # N.B.: We use a different version of requests from 3rdparty/python:requests because...
    requirement('requests==2.0.0'),
    'src/python/foo:bar'
  ]
)
```

A custom constraint is only needed to control the version of an interior node your code does not directly depend on.